src/tools/sc-hsm-tool.c

```diff
@@ -1667,6 +1693,208 @@ static int unwrap_key(sc_card_t *card, int keyid, const char *inf, const char *p
+	if (ef_cert_len < 0) {
+		fprintf(stderr, "Error reading certificate %s. Skipping\n", sc_strerror(ef_cert_len));
+		ef_cert_len = 0;
+	} else {
+		ef_cert_len = determineLength(ef_cert, ef_cert_len);
+	}
+
+	/* C_DevAut */
+	fid[0] = 0x2F;
+	fid[1] = 0x02;
+	dev_aut_cert_len = 0;
+
+	/* Read concatenation of both certificates */
+	sc_path_set(&path, SC_PATH_TYPE_FILE_ID, fid, sizeof(fid), 0, 0);
+	r = sc_select_file(card, &path, NULL);
+	if (r != SC_SUCCESS) {
+		fprintf(stderr, "Failed to select certificates: %s\n", sc_strerror(r));
+		return -1;
+	}
+
+	total_certs_len = sc_read_binary(card, 0, dev_aut_cert, sizeof(dev_aut_cert), 0);
+	if (total_certs_len < 0) {
+		fprintf(stderr, "Error reading certificate: %s\n", sc_strerror(total_certs_len));
+		return -1;
+	} else {
+		dev_aut_cert_len = determineLength(dev_aut_cert, total_certs_len);
+		dica_len = total_certs_len - dev_aut_cert_len;
+		memcpy(dica, dev_aut_cert + dev_aut_cert_len, dica_len);
+	}
+	if (dica_len == 0) {
+		fprintf(stderr, "Could not determine device issuer certificate\n");
+		return -1;
+	}
+
+	if ((outfp = fopen(outf, "r"))) {
+		fprintf(stderr, "Output file '%s' already exists\n", outf);
+		fclose(outfp);
+		return -1;
+	}
+	fprintf(stderr, "Warning: Device certificate chain not verified!\n");
+
+	datalen = ef_cert_len + dev_aut_cert_len + dica_len;
+	outlen = 8 + datalen;
+	if (!(data = malloc(datalen))) {
+		fprintf(stderr, "Malloc failed\n");
+		r = -1;
+		goto err;
+	}
+	if (!(out = malloc(outlen))) {
+		fprintf(stderr, "Malloc failed\n");
+		r = -1;
+		goto err;
+	}
+	memcpy(data, ef_cert, ef_cert_len);
+	memcpy(data + ef_cert_len, dev_aut_cert, dev_aut_cert_len);
+	memcpy(data + ef_cert_len + dev_aut_cert_len, dica, dica_len);
+
+	if ((r = sc_asn1_put_tag(tag, data, datalen, out, outlen, &ptr)) < 0) {
+		fprintf(stderr, "Error formatting ASN1 sequence: %s\n", sc_strerror(r));
+		r = -1;
+		goto err;
+	}
+	outlen = ptr - out;
+
+	if (!(outfp = fopen(outf, "wb"))) {
+		perror(outf);
+		r = -1;
+		goto err;
+	}
+
+	if (fwrite(out, 1, outlen, outfp) != (size_t)outlen) {
+		perror(outf);
+		r = -1;
+		goto err;
+	}
+
+err:
+	if (outfp)
+		fclose(outfp);
+	if (out)
+		free(out);
+	if (data)
+		free(data);
+
+	return r;
+}
+
+static int register_public_key(sc_context_t *ctx, sc_card_t *card, const char *inf)
+{
+	FILE *in = NULL;
+	struct stat sb;
+	u8 *pka = NULL;
+	int r = 0;
+	sc_cardctl_sc_hsm_pka_register_t pka_register;
+
+	if (!(in = fopen(inf, "rb"))) {
+		perror(inf);
+		r = -1;
+		goto err;
+	}
+	if (fstat(fileno(in), &sb)) {
+		perror("cannot fstat");
+		r = -1;
+		goto err;
+	}
+	if (sb.st_size == 0) {
+		fprintf(stderr, "File is empty\n");
+		r = -1;
+		goto err;
+	}
+	if (!(pka = malloc(sb.st_size))) {
+		fprintf(stderr, "Malloc failed\n");
+		r = -1;
+		goto err;
+	}
+	if (fread(pka, 1, sb.st_size, in) != (size_t)sb.st_size) {
+		perror(inf);
+		r = -1;
+		goto err;
+	}
+	fclose(in);
+	in = NULL;
```

Comment:
you may want to have a look at `fread_to_eof()`...

Reply:
thanks, replaced with `fread_to_eof`